### PR TITLE
Use all intermediate .gcov files that are generated.

### LINKED ISSUE #5880: Use gcov for collecting code coverage for gcc
Currently C++ coverage is 2x slower than just running the corresponding tests. The goal is to have similar times for running bazel test and bazel coverage on the same targets. When `gcc` is used `gcov` is the fastest tool for computing code coverage results.

### LINKED ISSUE #6129: Make C++ code coverage collection correct
Currently the results in the code coverage report generated by Bazel for C++ code are not respecting the `--instrumentation_filter` and `--instrument_test_targets` flags.

### PR DESCRIPTION
**Pull request by iirina**
Previously we would only use one `.gcov` file per `gcov` invocation for the final coverage report. This is incorrect because `gcov` generates multiple `.gcov` files for all the sources and headers used for creating the corresponding `.gcda`/`.gcno` files.

One example affected by this is when a `cc_library` has multiple headers. We need coverage information for all the headers used.

This PR adds several tests to make sure coverage collection works correctly when a `cc_library` has multiple header files and when a header file is used by multiple `cc_library` targets.

Unnecessary coverage information will be filtered out by `CoverageOutputGenerator`.

Fixes #6129 
Fixes #5880

### CODE REVIEW

**Comment by iirina** [comment]
@mhlopko ping :smiley:

File: tools/test/collect_cc_coverage.sh [hunk 1]
```diff
@@ -113,7 +113,8 @@ function lcov_coverage() {
 function gcov_coverage() {
   local output_file="${1}"; shift
 
-  touch "$output_file"
+  # We'll save the standard output of each the gcov command in this log.
+  local gcov_log="$output_file.gcov.log"
 
   # Move .gcno files in $COVERAGE_DIR as the gcda files, because gcov
```
Comment (mhlopko) [inline tools/test/collect_cc_coverage.sh]: Is "Copy .gcno files next to their corresponding gcda files in the $COVERAGE_DIR" more accurate? Did I get that right? :)

File: tools/test/collect_cc_coverage.sh [hunk 2]
```diff
@@ -113,7 +113,8 @@ function lcov_coverage() {
...
   # expects them to be under the same directory.
```
Comment (mhlopko) [inline tools/test/collect_cc_coverage.sh]: Below, "If the gcda file was not found we generate empty coverage from the gcno file.", where does it happen?

File: tools/test/collect_cc_coverage.sh [hunk 3]
```diff
@@ -140,92 +141,53 @@ function gcov_coverage() {
...
-            cat "$gcov_file" >> "${output_file}"
-            # We don't need this file anymore.
-            rm -f "$gcov_file"
-        fi
+        # (or header) file containing code which was compiled to produce the
+        # .gcda files.
+        "${GCOV}" -i -b -o "$(dirname ${gcda})" "${gcda}" &> "$gcov_log"
+
+        # Go through all the files that were created by the gcov command above
+        # and copy them in the output .gcov file.
```
Comment (mhlopko) [inline tools/test/collect_cc_coverage.sh]: Maybe "and append their content to the output .gcov file"?

File: src/test/shell/bazel/bazel_coverage_test.sh [hunk 4]
```diff
@@ -64,6 +64,25 @@ int main(void) {
 EOF
 }
 
+# Returns 0 if gcov is not installed or if a version before 7.0 was found.
```
Comment (mhlopko) [inline src/test/shell/bazel/bazel_coverage_test.sh]: I'd return 0 on success and error code instead

Comment (mhlopko) [reply src/test/shell/bazel/bazel_coverage_test.sh]: or rename the function to is_wrong_gcov_installed :)

File: src/test/shell/bazel/bazel_coverage_test.sh [hunk 5]
```diff
@@ -136,21 +156,10 @@ end_of_record"
 }
 
 function test_cc_test_coverage_gcov() {
-  local -r gcov_location=$(which gcov)
-  if [[ ! -x ${gcov_location:-/usr/bin/gcov} ]]; then
-    echo "gcov not installed. Skipping test."
-    return
+  if is_gcov_right_version_installed; then
+    echo "Skipping test." && return
```
Comment (mhlopko) [inline src/test/shell/bazel/bazel_coverage_test.sh]: This reads wrong, right? It should be if wrong_gcov_is_installed then echo "skipping test". Or am I off?

File: tools/test/collect_cc_coverage.sh (hunk 1 above)
Comment (iirina) [reply tools/test/collect_cc_coverage.sh]: Yup, thanks!

File: tools/test/collect_cc_coverage.sh (hunk 2 above)
Comment (iirina) [reply tools/test/collect_cc_coverage.sh]: I rephrased it - if the gcda is not found then we skip generating any coverage for this gcno.

File: src/test/shell/bazel/bazel_coverage_test.sh (hunk 5 above)
Comment (iirina) [reply src/test/shell/bazel/bazel_coverage_test.sh]: Pfff yes, I got so used to the code I totally missed it. Thanks!

File: src/test/shell/bazel/bazel_coverage_test.sh (hunk 4 above)
Comment (iirina) [reply src/test/shell/bazel/bazel_coverage_test.sh]: Renamed it.

**Comment by iirina** [comment]
@mhlopko ping

**Comment by iirina** [comment]
Thanks for the review!

**Pull request closed without merging**